Uplift `temporary-cstring-as-ptr` lint from `clippy` into rustc

nathanwhit opened the pull request:
The general consensus seems to be that this lint covers a common enough mistake to warrant inclusion in rustc. 
The diagnostic message might need some tweaking, as I'm not sure the use of second-person perspective matches the rest of rustc, but I'd like to hear others' thoughts on that.

(cc #53224).

r? @oli-obk

Mark-Simulacrum commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit f7da096488c6fd9d4be7c480affa26d9f7b80a02 with merge ca7ac508d257828878fb0e7cade1ad46381056f8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f7da096488c6fd9d4be7c480affa26d9f7b80a02","merge_sha":"ca7ac508d257828878fb0e7cade1ad46381056f8"} -->

oli-obk commented:
Sgtm, modulo CI passing, perf being clean and we'll need an [MCP](https://github.com/rust-lang/compiler-team/issues/new?assignees=&labels=major-change%2C+T-compiler&template=major_change.md&title=%28My+major+change+proposal%29) so the entire compiler team is informed about this. Can you open an MCP? It just needs a sentence or two and maybe an example on what is now being linted.

Mark-Simulacrum commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+        let method_names: Vec<SymbolStr> = method_names.iter().map(|s| s.as_str()).collect();
+        let method_names: Vec<&str> = method_names.iter().map(|s| &**s).collect();
+
+        if let ["as_ptr", "unwrap" | "expect"] = method_names.as_slice() {
```
Regardless of perf results, it seems pretty bad to allocate 3 vectors just to check if we ever see as_ptr followed by unwrap or expect. It seems like we can fairly easily do this without the multiple vectors allocated, though it would require some restructuring. It also looks like arg_lists doesn't need more than the first two elements, so we can avoid pushing after that as well.

nathanwhit replied:
Just pushed some changes to avoid all of the allocations.

Mark-Simulacrum commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+
+        let (method_names, arg_lists, _) = method_calls(expr, 2);
+        let method_names: Vec<SymbolStr> = method_names.iter().map(|s| s.as_str()).collect();
+        let method_names: Vec<&str> = method_names.iter().map(|s| &**s).collect();
```
This dance to get the symbols to &str is also pretty unfortunate, it should be easy to add as_ptr, unwrap, and expect as "known symbols" in src/librustc_span/symbol.rs and then compare directly against those there (or, following my other comment, later).

We should also do this elsewhere where we're comparing against strings right now in this file.

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+
+declare_lint! {
+    pub TEMPORARY_CSTRING_AS_PTR,
+    Deny,
```
We'll need a crater run to check the fallout from this (but it should still be deny by default unless we break some foundational crate, IMO).

oli-obk replied:
dependencies get their lints capped anyway, so it will only break building that crate directly.

tmiasko replied:
I feel that it shouldn't be deny by default as long as it considers `func(CString::new("...").unwrap().as_ptr());` to be erroneous (since in typical situations this would be a false positive).

matklad replied:
+1 for making this Warn, and not Deny (though I don't know if we have other similar deny-by-default lints). 

I think we also should add a regression test for `func(CString::new("...").unwrap().as_ptr());`. I wouldn't personally call it as  a false-positive though: it's not UB, but still sufficiently obscure style worth linting against.

nathanwhit replied:
I agree on the Warn by default, since the lint covers cases that aren't *incorrect*. That's reflected in the PR now.

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+    unwrap: &rustc_hir::Expr<'_>,
+) {
+    let source_type = cx.typeck_results().expr_ty(source);
+    if let ty::Adt(def, substs) = source_type.kind {
+        if cx.tcx.is_diagnostic_item(Symbol::intern("result_type"), def.did) {
+            if let ty::Adt(adt, _) = substs.type_at(0).kind {
+                let path = [
+                    sym::std,
+                    Symbol::intern("ffi"),
+                    Symbol::intern("c_str"),
+                    Symbol::intern("CString"),
+                ];
+                if cx.match_def_path(adt.did, &path) {
```
I feel like all of this could be written with early returns to reduce rightward drift:


```suggestion
    if let ty::Adt(def, substs) = source_type.kind {
        (def, substs)
    } else {
        return;
    }
    if !cx.tcx.is_diagnostic_item(Symbol::intern("result_type"), def.did) {
        return;
    }
    if let ty::Adt(adt, _) = substs.type_at(0).kind {
        let path = [
            sym::std,
            Symbol::intern("ffi"),
            Symbol::intern("c_str"),
            Symbol::intern("CString"),
        ];
        if !cx.match_def_path(adt.did, &path) {
            return;
        }
    }
```

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+                if cx.match_def_path(adt.did, &path) {
+                    cx.struct_span_lint(TEMPORARY_CSTRING_AS_PTR, expr.span, |diag| {
+                        let mut diag = diag
+                            .build("you are getting the inner pointer of a temporary `CString`");
```
```suggestion
                            .build("getting inner pointer of temporary `CString`");
```

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+                    Symbol::intern("CString"),
+                ];
+                if cx.match_def_path(adt.did, &path) {
+                    cx.struct_span_lint(TEMPORARY_CSTRING_AS_PTR, expr.span, |diag| {
```
```suggestion
                    cx.struct_span_lint(TEMPORARY_CSTRING_AS_PTR, source.span, |diag| {
```

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+                    cx.struct_span_lint(TEMPORARY_CSTRING_AS_PTR, expr.span, |diag| {
+                        let mut diag = diag
+                            .build("you are getting the inner pointer of a temporary `CString`");
+                        diag.note("that pointer will be invalid outside this expression");
```
```suggestion
                        diag.span_label(source.span, "this pointer will be invalid");
```

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+                        let mut diag = diag
+                            .build("you are getting the inner pointer of a temporary `CString`");
+                        diag.note("that pointer will be invalid outside this expression");
+                        diag.span_help(
+                            unwrap.span,
+                            "assign the `CString` to a variable to extend its lifetime",
+                        );
```
```suggestion
                        diag.span_help(
                            unwrap.span,
                            "this `CString` is deallocated at the end of the expression, bind it to variable to extend its lifetime",
                        );
```

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,102 @@
+                            unwrap.span,
+                            "assign the `CString` to a variable to extend its lifetime",
+                        );
+                        diag.emit();
```
```suggestion
                        diag.note("pointers do not have a lifetime; when calling `as_ptr` the `CString` is deallocated because nothing is referencing to it as far as the type system is concerned");
                        diag.emit();
```

nathanwhit commented:
> Sgtm, modulo CI passing, perf being clean and we'll need an [MCP](https://github.com/rust-lang/compiler-team/issues/new?assignees=&labels=major-change%2C+T-compiler&template=major_change.md&title=%28My+major+change+proposal%29) so the entire compiler team is informed about this. Can you open an MCP? It just needs a sentence or two and maybe an example on what is now being linted.

Sure thing! Filed rust-lang/compiler-team#346.

estebank commented on src/test/ui/lint/lint-temporary-cstring-as-ptr.stderr:
```diff
@@ -1,16 +1,16 @@
    |
 LL |     let s = CString::new("some text").unwrap().as_ptr();
-   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^ this pointer will be invalid
```
Oh, I thought that the span would point to `as_ptr()`...

nathanwhit replied:
Whoops. Yup that definitely makes more sense.

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,90 @@
+                        let mut diag = diag
+                            .build("getting the inner pointer of a temporary `CString`");
+                        diag.span_label(as_ptr_span, "this pointer will be invalid");
+                        diag.span_help(
+                            unwrap.span,
+                            "this `CString` is deallocated at the end of the expression, bind it to a variable to extend its lifetime",
+                        );
```
Last nitpick: could we make this a `span_label` as well? I think that the following would look a bit better:

```
error: getting the inner pointer of a temporary `CString`
  --> $DIR/lint-temporary-cstring-as-ptr.rs:6:48
   |
LL |     let s = CString::new("some text").unwrap().as_ptr();
   |             ---------------------------------- ^^^^^^ this pointer will be invalid
   |             |
   |             this `CString` is deallocated at the end of the expression, bind it to a variable to extend its lifetime
   |
   = note: `#[deny(temporary_cstring_as_ptr)]` on by default
   = note: pointers do not have a lifetime; when calling `as_ptr` the `CString` is deallocated because nothing is referencing it as far as the type system is concerned
```

lzutao commented on library/std/src/ffi/c_str.rs:
````diff
@@ -1261,7 +1261,7 @@ impl CStr {
     ///
     /// ```no_run
-    /// # #![allow(unused_must_use)]
+    /// # #![allow(unused_must_use)] #![allow(temporary_cstring_as_ptr)]
````
```suggestion
    /// # #![allow(unused_must_use, temporary_cstring_as_ptr)]
```

jyn514 commented:
I think https://github.com/rust-lang/rust-clippy/issues/5838 should be fixed before making this deny-by-default. Warn-by-default is fine if fixing the false positive is too difficult and we want to get this in soon.

jyn514 commented:
> it is false-positive for "this code is UB", as it is not UB. But it absolutely is a true positive for "this code is confusing, because you need to be a language lawyer to understand why it is not UB".

Then it should say that. Instead it [currently says](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1435587d1f41aa6d78868f9a3774059b)

>   = note: that pointer will be invalid outside this expression

Which is not true. It will be invalid outside the _statement_, which encompasses both the `.as_ptr()` and the function call.

jyn514 commented:
Also I think it would be nice for it to link to https://doc.rust-lang.org/reference/destructors.html, but I won't block on that.

jyn514 commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,90 @@
+                        diag.span_label(as_ptr_span, "this pointer will be invalid");
+                        diag.span_label(
+                            unwrap.span,
+                            "this `CString` is deallocated at the end of the expression, bind it to a variable to extend its lifetime",
```
https://github.com/rust-lang/rust/pull/75671#issuecomment-684902197
```suggestion
                            "this `CString` is deallocated at the end of the statement, bind it to a variable to extend its lifetime",
```

jyn514 reviewed (commented):
I think this should add a test case for https://github.com/rust-lang/rust-clippy/issues/5838. If the behavior there is intentional, then it should be tested; it shouldn't slip in by accident. If it's not intentional, the lint should be fixed.

jyn514 commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,90 @@
+                            unwrap.span,
+                            "this `CString` is deallocated at the end of the expression, bind it to a variable to extend its lifetime",
+                        );
+                        diag.note("pointers do not have a lifetime; when calling `as_ptr` the `CString` is deallocated because nothing is referencing it as far as the type system is concerned");
```
Calling `as_ptr()` does not deallocate the string.
```suggestion
                        diag.note("pointers do not have a lifetime; when calling `as_ptr` the `CString` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned");
```

jyn514 commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,90 @@
+                            "this `CString` is deallocated at the end of the expression, bind it to a variable to extend its lifetime",
+                        );
+                        diag.note("pointers do not have a lifetime; when calling `as_ptr` the `CString` is deallocated because nothing is referencing it as far as the type system is concerned");
+                        diag.emit();
```
```suggestion
                        diag.help("for more information, see https://doc.rust-lang.org/reference/destructors.html");
                        diag.emit();
```

estebank commented on src/librustc_lint/methods.rs:
```diff
@@ -0,0 +1,90 @@
+    if let ty::Adt(def, substs) = source_type.kind {
+        if cx.tcx.is_diagnostic_item(sym::result_type, def.did) {
+            if let ty::Adt(adt, _) = substs.type_at(0).kind {
+                if cx.match_def_path(adt.did, &CSTRING_PATH) {
```
I would change this to use `is_diagnostic_item` by annotating `struct CString` with `#[rustc_diagnostic_item = "cstring_type"]`.

Now that I'm looking at the code, there are a bunch of lints using `match_def_path` that would be better served with using diagnostic item instead. (In practice this doesn't make a difference, but path checking is more brittle.)

nathanwhit commented:
Addressed the review comments and rebased, this should be ready for re-review.

jyn514 commented on src/test/ui/lint/lint-temporary-cstring-as-param.rs:
```diff
@@ -0,0 +1,10 @@
+// ignore-tidy-linelength
```
Maybe add a comment that this isn't technically incorrect, but it's not immediately obvious?

oli-obk commented:
Can you also remove it from clippy in the same PR?

estebank commented on compiler/rustc_lint/src/methods.rs:
```diff
@@ -0,0 +1,87 @@
+                            unwrap.span,
+                            "this `CString` is deallocated at the end of the statement, bind it to a variable to extend its lifetime",
+                        );
+                        diag.note("pointers do not have a lifetime; when calling `as_ptr` the `CString` will be deallocated at the end of the statement...");
+                        diag.note("...because nothing is referencing it as far as the type system is concerned");
```
I think these two `note`s can be merged into a single sentence. We already rely on cli reflow of text for long lines.

nathanwhit commented:
> Can you also remove it from clippy in the same PR?

Done!

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 4cc16db2d39f18ea1c5edbb92553df65d375ba86 has been approved by `oli-obk`

<!-- @bors r=oli-obk 4cc16db2d39f18ea1c5edbb92553df65d375ba86 -->
<!-- homu: {"type":"Approved","sha":"4cc16db2d39f18ea1c5edbb92553df65d375ba86","approver":"oli-obk"} -->

est31 commented:
Doctest failure in rollup #77082: 
```
---- src/ffi/c_str.rs - ffi::c_str::CStr::as_ptr (line 1269) stdout ----
error: unknown lint: `temporary_cstring_as_ptr`
 --> src/ffi/c_str.rs:1269:27
  |
3 | #![allow(unused_must_use, temporary_cstring_as_ptr)]
  |                           ^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/ffi/c_str.rs:1267:9
  |
1 | #![deny(warnings)]
  |         ^^^^^^^^
  = note: `#[deny(unknown_lints)]` implied by `#[deny(warnings)]`

```

jyn514 commented:
That looks like doc-tests are being run with stage 0, not stage 1. Do you know why that would happne?

ecstatic-morse commented:
@bors r-

Failed in rollup #77082.

est31 commented:
@jyn514 idk it seems to run [this](https://github.com/rust-lang-ci/rust/runs/1152638341#step:23:66) command:

```
  Step 9/9 : ENV SCRIPT python3 ../x.py test --stage 0 --config /config/nopt-std-config.toml library/std   && python3 ../x.py --stage 2 test
```

I was just checking what my the rollup my PR was in was up to and found this error :).

pickfire commented on library/std/src/ffi/c_str.rs:
```diff
@@ -109,7 +109,9 @@ use crate::sys;
 /// documentation of `CString` before use, as improper ownership management
 /// of `CString` instances can lead to invalid memory accesses, memory leaks,
 /// and other memory errors.
+
```
Is this necessary?
```suggestion
```

pickfire commented on src/test/ui/lint/lint-temporary-cstring-as-param.rs:
```diff
@@ -0,0 +1,10 @@
+fn some_function(data: *const i8) {}
+
+fn main() {
+    some_function(CString::new("").unwrap().as_ptr()); //~ ERROR getting the inner pointer of a temporary `CString`
```
Why not just remove the ignore tidy linelength check and reduce this line length?
```suggestion
    some_function(CString::new("").unwrap().as_ptr());
    //~^ ERROR getting the inner pointer of a temporary `CString`
```

pickfire commented on src/test/ui/lint/lint-temporary-cstring-as-ptr.rs:
```diff
@@ -0,0 +1,9 @@
+// ignore-tidy-linelength
+// this program is not technically incorrect, but is an obscure enough style to be worth linting
+#![deny(temporary_cstring_as_ptr)]
+
+use std::ffi::CString;
+
+fn main() {
+    let s = CString::new("some text").unwrap().as_ptr(); //~ ERROR getting the inner pointer of a temporary `CString`
```
```suggestion
// this program is not technically incorrect, but is an obscure enough style to be worth linting
#![deny(temporary_cstring_as_ptr)]

use std::ffi::CString;

fn main() {
    let s = CString::new("some text").unwrap().as_ptr();
    //~^ ERROR getting the inner pointer of a temporary `CString`
```

bors commented:
:umbrella: The latest upstream changes (presumably #77517) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

JohnCSimon commented:
Ping from triage:
@nathanwhit - can you please address the merge conflicts and the new comments on this PR? thanks

nathanwhit commented:
Rebased and addressed comments. I'm not sure why this failed in rollup before, hopefully it's resolved now?
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

jyn514 commented:
@nathanwhit you can replicate the failure with `./x.py test --stage 0 library/std`. I think you need `cfg(not(bootstrap))` on the allow.

nathanwhit commented:
> @nathanwhit you can replicate the failure with `./x.py test --stage 0 library/std`. I think you need `cfg(not(bootstrap))` on the allow.

Ah that indeed fixes it! Thanks for the help!
Should be ready for re-review/re-approval now.

jyn514 commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit 39941e6281e8da26ab6dfa776ff9583ec94cba71 has been approved by `oli-obk`

<!-- @bors r=oli-obk 39941e6281e8da26ab6dfa776ff9583ec94cba71 -->
<!-- homu: {"type":"Approved","sha":"39941e6281e8da26ab6dfa776ff9583ec94cba71","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 39941e6281e8da26ab6dfa776ff9583ec94cba71 with merge a4fea98f5bb6806d7840d91b4fe304873bcd0d0d...
<!-- homu: {"type":"BuildStarted","head_sha":"39941e6281e8da26ab6dfa776ff9583ec94cba71","merge_sha":"a4fea98f5bb6806d7840d91b4fe304873bcd0d0d"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1313249374)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1313249374","builder_name":"checks-actions"} -->

jyn514 commented:
This is on Android.
```
stderr:
------------------------------------------
error[E0308]: mismatched types
  --> /checkout/src/test/ui/lint/lint-temporary-cstring-as-param.rs:8:19
   |
LL |     some_function(CString::new("").unwrap().as_ptr());
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `i8`, found `u8`
   |
   = note: expected raw pointer `*const i8`
              found raw pointer `*const u8`

error: aborting due to previous error
```
I think `some_function` needs to take c_char instead of u8?

nathanwhit commented:
> This is on Android.
> 
> ```
> stderr:
> ------------------------------------------
> error[E0308]: mismatched types
>   --> /checkout/src/test/ui/lint/lint-temporary-cstring-as-param.rs:8:19
>    |
> LL |     some_function(CString::new("").unwrap().as_ptr());
>    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `i8`, found `u8`
>    |
>    = note: expected raw pointer `*const i8`
>               found raw pointer `*const u8`
> 
> error: aborting due to previous error
> ```
> 
> I think `some_function` needs to take c_char instead of u8?

Oops you're totally right. Should be fixed now.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 572cd358d379e4e641ad44bc693cdf48ceaf3581 has been approved by `oli-obk`

<!-- @bors r=oli-obk 572cd358d379e4e641ad44bc693cdf48ceaf3581 -->
<!-- homu: {"type":"Approved","sha":"572cd358d379e4e641ad44bc693cdf48ceaf3581","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 572cd358d379e4e641ad44bc693cdf48ceaf3581 with merge 902a3bc0467400209ed2e2b541136ebef3c494ce...
<!-- homu: {"type":"BuildStarted","head_sha":"572cd358d379e4e641ad44bc693cdf48ceaf3581","merge_sha":"902a3bc0467400209ed2e2b541136ebef3c494ce"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1315973218)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1315973218","builder_name":"checks-actions"} -->

jyn514 commented:
```
 Command failed. Attempt 5/5:
    Updating crates.io index
warning: spurious network error (2 tries remaining): error inflating zlib stream; class=Zlib (5)
warning: spurious network error (1 tries remaining): error inflating zlib stream; class=Zlib (5)
error: failed to get `cc` as a dependency of package `bootstrap v0.0.0 (D:\a\rust\rust\src\bootstrap)`

Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`

Caused by:
  error inflating zlib stream; class=Zlib (5)
failed to run: D:\a\rust\rust\build\i686-pc-windows-gnu\stage0\bin\cargo.exe build --manifest-path D:\a\rust\rust\src/bootstrap/Cargo.toml --locked
```

@bors retry

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1318298618)
Approved by: oli-obk
Pushing 90e6d0d46bae511aecea04ac5eeb3c387f0c04f4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1318298618"},"merge_sha":"90e6d0d46bae511aecea04ac5eeb3c387f0c04f4"} -->

flip1995 commented:
Uh, can we have a policy that the Clippy team is pinged for major changes to Clippy in rustc, such as uplifting lints. Even though the lint was removed, it wasn't deprecated properly (which is a lengthy process by design). Luckily @ebroto caught that in the rustup. But with a ping to the Clippy team we can avoid pull requests, like this "emergency sync": #78505
